Mathlib/Tactic/Linter/TextBased.lean

```diff
@@ -249,8 +260,9 @@ def formatErrors (errors : Array ErrorContext) (style : ErrorFormat) : IO Unit :
     IO.println (outputMessage e style)
 
 /-- Core logic of a text based linter: given a collection of lines,
-return an array of all style errors with line numbers. -/
-abbrev TextbasedLinter := Array String → Array (StyleError × ℕ)
+return an array of all style errors with line numbers and, optionally, an array of lines
+changed to fix the linter automatically. (This is not always possible, but sometimes it is.) -/
+abbrev TextbasedLinter := Array String → Array (StyleError × ℕ) × (Option (Array String))
```

Comment:
```suggestion
abbrev TextbasedLinter := Array String → Array (StyleError × ℕ) × Array String
```
Would this work?  I tried it out and I do not think that there are any problems interpreting `#[]` as no change.  Although maybe I am missing something and you really want to distinguish between `none` and `some #[]` -- I could not see where, though.

This entails further changes below, replacing `none`s by `#[]`, `if let some c := changed` with `if !changed.isEmpty` and similar small details.

Reply:
Yes, this would work. Conceptually, this is slightly wrong (as `some #[]` would mean "I fixed your file with no lines" and `none` would mean "nothing to fix"); in practice, this would not hurt. I like using the type system, and `Option` encodes what I want. Using sentinel values (Python's `-1` for "not found", etc.) is less clear, in my opinion.

Would this change make the code easier to follow for you? That could be a good reason to make it.

Reply:
My suggestion was more inspired by the fact that I would actually encode the replacements with something closer to `HashMap Nat String`, where instead of carrying around the whole "replaced" array of lines, you just record the index of the lines to change and what the new value is (possibly recording also the type of error together with either the index or the changed line).

Reply:
I considered this design as well, but I'm not sure if that is actually simpler. After all, there will be several auto-fixable linters in the long run, which should be chained. See [this comment](https://github.com/leanprover-community/mathlib4/pull/16332#discussion_r1743311849) for some details.